### ISSUE TITLE
How can codemeta increase discoverability by Google?

### ISSUE THREAD

**Issue by cboettig** [codemeta/codemeta#179]
(copying this question over from another issue thread by @maelle.  We ought to get this hashed out a bit better and then could maybe address this directly on https://codemeta.github.io)

Google uses and encourages the use of JSON-LD markup using the schema.org terms: https://developers.google.com/search/docs/guides/intro-structured-data .  This focuses on 'rich card' results shown on primarily on mobile devices, but applies to search generally.

At the moment, just adding a `codemeta.json` file to a github repo probably won't do anything to Google search since it's not embedded into an html file.  If we embedded into, say, html version of the README page CRAN generates, or into a pkgdown-generated site for a package, since then it would be crawled by Google.  There's currently no rich card format for software, but the use of rich markup may still increase search rank.  

(Google aside, the hope is also that this improves search in specific repositories that adopt or learn to read codemeta,  such as Datacite: https://search.datacite.org/works?query=Software  or http://citeas.org/ , Software Heritage, Zenodo, DataONE).

**Comment by maelle** [comment]
Reg _may_ could one test the effect or ask someone at Google?

**Comment by mbjones** [comment]
We are about to embed schema.org metadata in landing pages in DataONE, and that would include CodeMeta if we have it for a package. One challenge is that Google enforces strict type checking through the [Google Structured Types Testing Tool](https://search.google.com/structured-data/testing-tool/u/0/), and doesn't allow direct type references to classes outside of schema.org and a few other vocabularies.  To pass the Google tester with other vocabularies, the links must be embedded with the [schema:additionalType](https://schema.org/additionalType) property.  So we need to figure out how to serialize that with our CodeMeta extension properties.  You can see a more detailed discussion on how several data repositories are working to embed schema.org in data and software packages on the EarthCube Project418 repository:  https://github.com/earthcubearchitecture-project418/p418Vocabulary  -- I'm hoping DataONE will have that working for all of our data landing pages in Q1 of 2018.

**Comment by cboettig** [comment]
Thanks @mbjones that's awesome.  

Right, the non-schema terms create an issue.  I don't quite follow how the EarthCube examples are getting around this, the [example](https://raw.githubusercontent.com/earthcubearchitecture-project418/p418Vocabulary/master/html/voc/static/schema/examples/resource/dataset-full.jsonld) I looked at doesn't seem to use any properties from any vocabulary other than schema.org (I see that prefixes for other vocabularies are defined, but the validator doesn't seem to care what string you put into `additionalType`).  

Google's Testing Tool has been very opaque to me: mostly it is following the definitions I can see at https://schema.org for the specified type, but I have no idea where it is getting the list of required vs recommended or optional properties.  (Also, some fields have controlled vocabularies that are also not obvious to me -- for instance `contactType` from the official example can be "customer service", but any other string I've tried is not valid.)  Presumably they have some json schema as well?

**Comment by mbjones** [comment]
@cboettig The example you cited has a field `"additionalType": "geolink:Person"`, that's an example how we are adding in other vocabularies.

**Comment by cboettig** [comment]
Right, I saw that, just wasn't sure in what sense it was really adding the vocabulary. i.e. if `geolink:Person` type defines some additional property `x`, we still cannot add said property to the object by adding a `"geolink:x": "some value"`; only those properties defined under the official schema.org Types are valid.  (I think that was your original point as well, I just didn't follow it).  

Based on the schema.org description: 

>  An additional type for the item, typically used for adding more specific types from external vocabularies in microdata syntax. This is a relationship between something and a class that the thing is in. In RDFa syntax, it is better to use the native RDFa syntax - the 'typeof' attribute - for multiple types. Schema.org tools may have only weaker understanding of extra types, in particular those defined externally.

it looks like `additionalType` was introduced only because microdata doesn't permit vector types, (presumably this blurb pre-dates JSON-LD serialization), and the Google validator doesn't complain about un-recognized types because it has only a "weaker understanding"; i.e. only cares that you have a valid URI in this field.  I think that means that as far as meeting Google Validation we're stuck with schema.org properties and I don't see much hope for getting the additional codemeta terms included (short of getting them accepted into schema.org, as @mfenner has suggested we consider doing).  

It's not obvious that Google would ignore all structured data that included additional terms, rather than just consuming the terms it understood, but I guess they may be choosing to ignore the whole thing in that case; (even though it's perfectly valid JSON-LD and JSON-LD tools are built explicitly so that you can add additional properties without breaking things... oof walled gardens)

**Comment by mfenner** [comment]
In the short term I wouldn't worry too much about the Google Validator. Medium term it would be great if codemeta is fully aligned with schema.org (by changes on both ends), not sure how much extra effort is needed.

Since GitHub (and I guess other code repositories) don't allow embedding of HTML, this really only works in other places. Would be good to explore CRAN, npm, PyPi and Rubygems in this regard, if it hasn't been done already.

**Comment by cboettig** [comment]
Thanks @mfenner , this sounds like a good direction to go.

Currently codemeta has 10 terms that are not in schema.org: https://codemeta.github.io/terms/ .
A handful of those are arguably overly specific anyhow and difficult to generalize across software projects. The `funding` element is under discussion in #160.  

It may be worth noting that additionally, codemeta includes some terms schema.org terms from [`SoftwareApplication`](http://schema.org/SoftwareApplication), which the Google validator at least is not happy seeing attached to the `SoftwareSourceCode` object type.  

Codemeta currently also excludes certain terms that are valid under schema.org and even google validator (though mostly but not always ones quite irrelevant to software).  Issue #161 proposes changing this so that codemeta would only extend schema.org and not exclude terms.  This would ensure any valid schema.org `SoftwareSourceCode` was also a valid codemeta file (which will almost always be true but not always be true).  

I like the idea of embedding this kind of information into CRAN, npm, PyPi sites; haven't checked all of them but haven't seen much of this so far.  For starters I'll explore CRAN via user-side here: https://github.com/ropensci/codemetar/issues/46

**Comment by danbri** [comment]
Hi folks. I work on Schema.org, at Google. Great to see all this work, can you help me understand how far you've got and what the current status of the schema design looks like especially with regard to the Schema.org relationship? I've some catching up to do! Are there any tweaks to schema.org to suggest, or you're happy doing everything in codemeta? Also noticed https://github.com/codemeta/codemeta/issues/103 https://github.com/codemeta/codemeta/issues/134 ...

**Comment by cboettig** [comment]
@danbri Thanks for getting in touch!  A few quick answers:

- Most terms in codemeta come directly from Schema.org, while currently ten additional terms are provided by CodeMeta; see the table here:  https://codemeta.github.io/terms/

- CodeMeta use several terms from `schema:SoftwareApplication` (as well as `schema:SoftwareSourceCode`) when describing a `codemeta:SoftwareSourceCode` object.  We understand the distinction between source code and the actual application, but this distinction can be unhelpful in describing research software which is often installed from source to begin with.

- Codemeta currently only include those schema.org terms which it explicitly states in the context file (see https://codemeta.github.io/terms/ list); e.g. other terms from `CreativeWork` that clearly don't apply to software are not valid codemeta.  However, we are evaluating if this is unnecessarily strict: https://github.com/codemeta/codemeta/issues/161

- `author` is an (ordered) list in schema.org but an (unordered) set in schema.org (https://github.com/codemeta/codemeta/blob/master/codemeta.jsonld#L67)

- CodeMeta is discussing how to represent research funding metadata: https://github.com/codemeta/codemeta/issues/160 . (Other related groups are also interested in this, as linked in that thread; including one current open issue on schemaorg GitHub).

We would welcome your feedback on any of the above issues.

Thanks also for raising the question of possible tweaks to Schema.org; that is something we have discussed as well.  

- I think we would be very interested in proposing a funding type to schema.org; I believe as a community we are close to but have not yet fully converged on what we'd like that to look like.
  
- I  think we would be interested in proposing both some (but perhaps not all) of the 10 additional codemeta fields for inclusion in Schema.org's `SoftwareSourceCode` type definition.  Again, your thoughts or questions on the use case for these terms would be appreciated.  (It might be worth comparing against the [crosswalk table](https://codemeta.github.io/crosswalk/) to see how many of our currently crosswalked vocabularies are actually making use of these extended terms).  

It would be great to get your input on any possible proposal to Schema.org before we open an issue over at https://github.com/schemaorg/schemaorg.  

Cheers!

**Comment by danbri** [comment]
Thanks for the summary! Have you looked at SHACL or SHEX (RDF validation languages, see http://book.validatingrdf.com) to capture something of the more domain-specific structures you are defining?

**Comment by alee** [comment]
I finally added codemeta support to our repository at comses.net/codebases as part of the recent Scientific Software Registries Collaboration Workshop. I've embedded the codemeta in our codebase landing pages as embedded `application/ld+json` data but as expected from this thread and testing via [Google's structured data testing tool](https://search.google.com/structured-data/testing-tool) I'm starting to get complaints in our search console about unparsable structured data. 

I think the main culprit is the `@context` being set to `["https://doi.org/10.5063/schema/codemeta-2.0", "http://schema.org"]` - the terms that belonged to `softwareApplication` I added to `targetProduct` as @dgarijo suggested during the workshop.

For an example of what the codemeta looks like you can view source on any model in the library (e.g., https://doi.org/10.25937/krp4-g724)

In this case should I just use `@context: "http://schema.org"`?

**Comment by cboettig** [comment]
@danbri can probably elaborate further, but I believe Google's structured data testing tool is not intended to be a generic validator of the W3C JSON-LD specification, but rather, whether the JSON-LD conforms to Google's specific use cases.  

That said, I think just going with `@context: "http://schema.org"` makes sense to me in this case.  It doesn't look like you have any terms that aren't already `schema.org` terms, so I think either way should be considered to be a valid codemeta file.  Furthering the already close alignment between CodeMeta and schema.org vocabularies would improve interoperability overall anyway.

**Comment by danbri** [comment]
Yes, what Carl says: it checks that Google can understand the dialects of
json-ld understood by 25-30 or so Google product features

**Comment by maelle** [comment]
So embedding a codemeta.json as is doesn't hurt and could even help, correct?

Does the codemeta documentation state this and the validation problem mentioned above?

**Comment by alee** [comment]
We're hopeful that it'll improve Google search results / discovery of our landing pages for archived computational models..

I'd be happy to add a PR to the codemeta documentation site if it seems useful. The developer guide describes the context here: https://codemeta.github.io/developer-guide/#extending-the-codemeta-context - is this where changes would be recommended wrt embedding codemeta as structured data in a webpage?

**Comment by rebeccaringuette** [comment]
@dgarijo and I have hopefully made significant progress towards this by creating a codemeta-compliant json-ld file that both validates in Schema.org and produces no critical issues for the indicated software item using Google's RichResults tool. Please take a look [Merged PR](https://github.com/TheOSSci/InnovationSprint2025_Extending_Software_Metadata/pull/7) and [example json-ld](https://github.com/TheOSSci/InnovationSprint2025_Extending_Software_Metadata/blob/main/schema.org_codemeta-compliant.json). The idea is that this type of file could be embedded in websites to improve alignment with those tools in a codemeta approved way. Once CodeMeta accepts this (if this hasn't happened already), we can work with Zenodo, GitHub, and possibly Software Heritage to create a pipeline where such metadata can be embedded on those landing pages for the software item in question. I imagine some question or input added to the forms or processes where the person can upload the codemeta-compliant schema.org json-ld file and the metadata gets embedded in the page.

**Comment by rebeccaringuette** [comment]
related to https://github.com/codemeta/codemeta/issues/446
Google's schema.org validator will import the definitions of terms if available as in issue 446.
